components/identity-mgt/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/listener/TenantSyncListener.java

```
@@ -0,0 +1,401 @@
+            LOG.debug("Checking if tenant can be created. Tenant ID: " + tenantId);
+        }
+
+        boolean canCreateTenant = false;
+
+        IdentityEventListenerConfig identityEventListenerConfig = IdentityUtil.readEventListenerProperty(
+                UserOperationEventListener.class.getName(), TenantSyncListener.class.getName());
+
+        if (StringUtils.isNotBlank(identityEventListenerConfig.getEnable())
+                && Boolean.parseBoolean(identityEventListenerConfig.getEnable())) {
+            try {
+                PrivilegedCarbonContext.getThreadLocalCarbonContext().startTenantFlow();
+                PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantId(tenantId);
+
+                RealmService realmService = IdentityMgtServiceComponent.getRealmService();
+
+                Tenant tenant = realmService.getTenantManager().getTenant(tenantId);
+                String organizationID = tenant.getAssociatedOrganizationUUID();
+
+                // check if the Organization Depth in the Hierarchy is 0. only then create the root org.
+                if (organizationID == null || IdentityMgtServiceDataHolder.getInstance().getOrganizationManager()
+                        .getOrganizationDepthInHierarchy(organizationID) == 0) {
+
+                    canCreateTenant = true;
+
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debug("Tenant is a root organization and can be created. Tenant ID: "
+                                + tenantId);
+                    }
+
+                } else {
+                    LOG.debug("Skipping creating the tenant since the triggered Event is not related "
+                            + "to a root org creation.");
+                }
+
+                // if there was an exception thrown here, tenant activation won't happen
+            } catch (UserStoreException | OrganizationManagementServerException e) {
+                LOG.error("Error while creating tenant ", e);
```

It's better to throw an exception and handle it in one place rather than logging it everywhere.

---

Didn't throw an exception since we do not want to stop the process from continuing.  only want to log the error and continue. Hope this is fine